Added toggle service

rmkraus opened the pull request:
Added a toggle service to the homeassistant, light, switch, and
media_player domains.

rmkraus committed:
Added toggle service

Added a toggle service to the homeassistant, light, switch, and
media_player domains.

balloob commented on homeassistant/components/light/__init__.py:
```diff
@@ -165,9 +178,15 @@ def handle_light_service(service):
             params[ATTR_TRANSITION] = transition
 
+        service_fun = None
         if service.service == SERVICE_TURN_OFF:
```
Wouldn't you be able to write this as?
```python
if service.service in (SERVICE_TUNR_OFF, SERVICE_TOGGLE):
    for light in target_lights:
        getattr(light, service.service)(**params)
```

rmkraus replied:
Yes and no. This would break if we ever changed the value of SERVICE_TOGGLE or SERVICE_TURN_OFF. This probably won't happen soon, but it could happen if we ever wanted to consider internationalization. I have no problem either way as it would be an easy fix when and if that ever happened. It's up to you, really.

balloob replied:
Yeah you're right, we shouldn't couple service names and attribute names.

Thought of 1 more optimization.
```python
if service.service == SERVICE_TURN_OFF:
    service_fun = light.turn_off
[…]

if service_fun:
    service_fun(**params)
```

balloob commented on homeassistant/helpers/entity.py:
```diff
@@ -174,3 +174,10 @@ def turn_on(self, **kwargs):
+
+    def toggle(self, **kwargs):
+        """ Toggle the entity off. """
+        if self.state == STATE_ON:
```
Please use the `is_on` property here.

balloob commented:
Can you add some tests too

rmkraus committed:
Added tests for toggle service.

1) Added tests to toggle service.
2) Removed color_util import in light tests. It was not being used.

rmkraus committed:
Revised entity toggle to use is_on

The toggle function in the Entity ABC was using state == STATE_ON to
determine whether the entity was on. This was revised to use the is_on
property instead.

balloob commented:
:dolphin:
